Code/GraphMol/RGroupDecomposition/RGroupScore.cpp

```diff
@@ -48,64 +43,40 @@ double score(const std::vector<size_t> &permutation,
         std::cerr << " RGroup: " << rg->second->smiles << " "
                   << rg->second->is_hydrogen << std::endl;;
 #endif
-	rgroup_column &col = matchSet[rg->second->smiles];
-	col.count += 1;
-        // detect whether or not this is an H
-        if (rg->second->is_hydrogen) {
-	  col.onlyH = true;
-        } else {
-	  col.onlyH = false;
-        }
+        unsigned int &count = matchSet[rg->second->smiles];
+        ++count;
 #ifdef DEBUG
         std::cerr << " " << rg->second->combinedMol->getNumAtoms(false)
-                  << " isH: " << onlyH[rg->second->smiles]
-                  << " score: " << matchSet[rg->second->smiles] << std::endl;
+                  << " score: " << count << std::endl;
 #endif
         if (rg->second->is_linker) {
-          linkerMatchSet[rg->second->attachments]++;
+          ++linkerMatchSet[rg->second->attachments];
 #ifdef DEBUG
           std::cerr << " Linker Score: "
-                    << linkerMatchSet[rg->second->attachments]++ << std::endl;
+                    << linkerMatchSet[rg->second->attachments] << std::endl;
 #endif
         }
       }
     }
     
     // get the counts for each rgroup found and sort in reverse order
-    std::vector<float> equivalentRGroupCount;
-
-    for (const auto &kv :  matchSet) {
-#ifdef DEBUG
-      std::cerr << " equiv: " << it->first << " " << it->second << " "
-                << permutation.size() << std::endl;
-#endif
+    std::vector<unsigned int> equivalentRGroupCount;
 
-      // if the rgroup is hydrogens, only consider if the group is all
-      //  hydrogen, otherwise score based on the non hydrogens
-      if (kv.second.onlyH) {
-        if (static_cast<size_t>(kv.second.count) == permutation.size()) {
-          equivalentRGroupCount.push_back(static_cast<float>(kv.second.count));
-        } else {
-          // hydrogens in a mixed group don't contribute to the score
-          equivalentRGroupCount.push_back(0.0);
-        }
-      } else {
-        equivalentRGroupCount.push_back(static_cast<float>(kv.second.count));
-      }
-    }
+    std::transform(
+        matchSet.begin(), matchSet.end(),
+        std::back_inserter(equivalentRGroupCount),
+        [](const std::pair<std::string, unsigned int> &p) { return p.second; });
     std::sort(equivalentRGroupCount.begin(), equivalentRGroupCount.end(),
-              std::greater<float>());
+              std::greater<unsigned int>());
 
-    double tempScore = 1.;
+    double tempScore = 0.;
     // score the sets from the largest to the smallest
     //  each smaller set gets penalized (i+1) below
     //  1.0 is the perfect score
     for (size_t i = 0; i < equivalentRGroupCount.size(); ++i) {
-      auto lscore =
-          equivalentRGroupCount[i] / ((i + 1) * (double)matches.size());
-      if (lscore > 0) {
-        tempScore *= lscore * lscore;
-      }
```

Comment:
These two codes do vastly different things.  (*= versus +=) I'll have to look at my notes for why we switched from += to *=, but they are old.

Reply:
I do realize that, but none of the unit tests breaks, and the tests that I have added are now scored as they should whereas the previous scoring function did not get them right.

Take the case where `R1` has 3 `-Cl` and 1 `-F`, and `R2` has 3 `-F` and 1 `-Cl` which should be preferred to the case where both `R1` and `R2` have 2 `-Cl` and 2 `-F`.
The original scoring function yields the following scores:
`R1` has 3 `-Cl` and 1 `-F`: score = (3.0 / (1.0 * 4.0))^2 * (1.0 / (2.0 * 4.0))^2 = 0.008789
`R1` has 2 `-Cl` and 2 `-F`: score = (2.0 / (1.0 * 4.0))^2 * (2.0 / (2.0 * 4.0))^2 = 0.015625
So the second incorrectly scores higher.

The new scoring function yields the following scores:
`R1` has 3 `-Cl` and 1 `-F`: score = (3.0 / (1.0 * 4.0))^2 + (1.0 / (2.0 * 4.0))^2 = 0.578125
`R1` has 2 `-Cl` and 2 `-F`: score = (2.0 / (1.0 * 4.0))^2 + (2.0 / (2.0 * 4.0))^2 = 0.3125
So the first correctly scores higher.

Reply:
The point that the changes here don't break any of the existing tests is a really strong one.
(It's also probably a good indication that our existing tests are not as comprehensive as they should be... we ought to work on that in some future PR).

Reply:
@ptosco - the original code had += and I'm happy with the change.  I'm doing a last pass with a synthetic benchmark tonight, I'm taking a highly symmetric core like naphthalene and adding known #'s of groups and filling in the rest, this is really a formality to check for any future failures and will help me in the future when I add threaded evaluations and the rooted-atom scoring function :)

I like the rest of the changes.  Expect this to be merged tomorrow.